subsys/net/lib/tls_credentials/tls_credentials_trusted.c

```diff
@@ -456,3 +460,399 @@ int tls_credential_delete(sec_tag_t tag, enum tls_credential_type type)
+	/* PSA keygen */
+	psa_status_t psa_status;
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+	psa_key_id_t key_id = PSA_KEY_ID_NULL;
+	psa_key_usage_t key_usage;
+
+	/* MbedTLS key formatting */
+	int mbed_status;
+	mbedtls_pk_context pk_ctx;
+	void* key_start = 0;
+	size_t key_max;
+
+	/* Permanent key storage */
+	int cred_status;
+	psa_storage_uid_t uid = tls_credential_get_uid(tag, TLS_CREDENTIAL_PRIVATE_KEY);
+	bool key_stored = false;
+
+	if (key_len == NULL) {
+		return -EINVAL;
+	}
+
+	key_max = *key_len;
+
+	k_mutex_lock(&credential_lock, K_FOREVER);
+	mbedtls_pk_init(&pk_ctx);
+
+	/* Ensure the storage destination is valid. */
+	if (!tag_type_valid(tag, TLS_CREDENTIAL_PRIVATE_KEY)) {
+		ret = -EINVAL;
+		goto cleanup;
+	}
+
+	/* Presently, only SECP256R1 is supported by this backend for keygen */
+	if (!tls_credential_can_keygen(type)) {
+		ret = -ENOTSUP;
+		goto cleanup;
+	}
+
+	/* Verify sectag not already taken */
+	if (tls_credential_toc_find_slot(uid) != CRED_MAX_SLOTS) {
+		ret = -EEXIST;
+		goto cleanup;
+	}
+
+	/* Before attempting keygen, check that an empty slot is available */
+	if (tls_credential_toc_find_slot(0) == CRED_MAX_SLOTS) {
+		ret = -ENOMEM;
+		goto cleanup;
+	}
+
+	/* Use PSA to generate a volatile SECP256R1 private/public key-pair
+         *
+	 * It is possible (and better) to mark the private key as persistent, and store it directly
+	 * in Trusted Internal Storage. This would be more secure, but until we support opaque
+	 * keys properly, it isn't an option.
+	 *
+	 * Accordingly, for now, we simply generate a volatile key and transfer it into
+	 * protected storage.
+	 */
+
+	key_usage = PSA_KEY_USAGE_EXPORT;
+	psa_set_key_usage_flags(&key_attributes, key_usage);
+
+	psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+	psa_set_key_algorithm(&key_attributes, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+	psa_set_key_type(&key_attributes, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+	psa_set_key_bits(&key_attributes, 256);
+
+	psa_status = psa_crypto_init();
+
+	if (psa_status != PSA_SUCCESS) {
+		LOG_ERR("Failed to initialize crypto. Status: %d", psa_status);
+		ret = -EFAULT;
+		goto cleanup;
+	}
+
+	psa_status = psa_generate_key(&key_attributes, &key_id);
+
+	if (psa_status != PSA_SUCCESS) {
+		LOG_ERR("Failed to generate private key. Status: %d", psa_status);
+		ret = -EFAULT;
+		goto cleanup;
+	}
+
+	/* Hand key off to MbedTLS for immediate use (CSR and formatted export). */
+	mbed_status = mbedtls_pk_setup_opaque(&pk_ctx, key_id);
+
+	if (mbed_status) {
+		LOG_ERR("Failed to set up opaque private key. Status: %d", mbed_status);
+		ret = -EFAULT;
+		goto cleanup;
+	}
+
+	/* Reset key buffer before using it */
+	memset(key_buf, 0, key_max);
+	*key_len = 0;
+
+	/* Export private key material in the same RFC5915/SEC1 DER format that
+	 * MbedTLS will later expect when loading the private key material from storage.
+	 *
+	 * Temporarily use the public key output buffer for this.
```

Comment:
Mostly I did things this way out of convenience -- It is a nice way to avoid permanent stack costs and also to avoid hard-coding a maximum key-size. 

In later versions, key handling should be entirely opaque, so we won't need private key buffers at all. So either way what we do here isn't going to be permanent.

Maybe let's see what Aegir's opinion is on this? @tomi-font What are your thoughts?